File: services/agent-engine/app/api/openclaw.py
"""OpenClaw control endpoints — detect, install, configure, operate."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlmodel import Session

from ..core.config import Settings
from ..db.session import get_session

from ..schemas.openclaw import (
    AddChannelRequest,
    AgentActionResult,
    AgentChatRequest,
    AgentChatResult,
    AgentDiscordResult,
    AgentDiscordStatus,
    AgentDocumentContent,
    AgentDocumentsResponse,
    AgentFilesResponse,
    AllChatsResponse,
    ConnectDiscordRequest,
    ChannelActionResult,
    ChannelsResponse,
    ClearSessionsRequest,
    ClearSessionsResult,
    CreateOpenClawAgentRequest,
    FileHandoffRequest,
    FileHandoffResult,
    DashboardUrlResponse,
    GatewayActionResult,
    OpenClawAgentResult,
    OpenClawAgentsResponse,
    OpenClawConfigRequest,
    OpenClawConfigResult,
    OpenClawInstallStatus,
    OpenClawStatusResponse,
    SaveAgentFileRequest,
    SessionDetail,
    SessionsResponse,
    WebSearchResult,
)
from ..services.activity_service import track
from ..services.chroma_service import ChromaService
from ..services.knowledge_service import KnowledgeService
from ..services.openclaw_service import OpenClawService
from .deps import get_app_settings

router = APIRouter(prefix="/openclaw", tags=["openclaw"])


def get_openclaw_service() -> OpenClawService:
    return OpenClawService()


@router.get("/status", response_model=OpenClawStatusResponse, summary="OpenClaw + prereqs status")
def openclaw_status(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.status()


@router.post("/install", response_model=OpenClawInstallStatus, summary="Install OpenClaw (npm -g)")
def openclaw_install(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.start_install()


@router.get("/install/status", response_model=OpenClawInstallStatus, summary="Install progress")
def openclaw_install_status(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.install_status()


@router.post("/config", response_model=OpenClawConfigResult, summary="Configure OpenClaw")
def openclaw_config(
    payload: OpenClawConfigRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Configure OpenClaw via its own non-interactive onboarding: free (Ollama /
    Gemma 4, no key) or a paid API key. Writes ~/.openclaw/openclaw.json."""
    return service.write_config(
        mode=payload.mode, model=payload.model, api_key=payload.api_key,
        provider=payload.provider, base_url=payload.base_url,
    )


@router.post("/gateway/start", response_model=GatewayActionResult, summary="Start the gateway")
def openclaw_gateway_start(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.gateway_start()


@router.post("/gateway/stop", response_model=GatewayActionResult, summary="Stop the gateway")
def openclaw_gateway_stop(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.gateway_stop()


@router.get("/dashboard", response_model=DashboardUrlResponse, summary="Get dashboard URL")
def openclaw_dashboard(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    """Return the token-authenticated OpenClaw dashboard URL (starts the gateway
    if needed). The desktop embeds it in a panel so clients never leave Evano."""
    return service.dashboard_url()


@router.post("/file-handoff", response_model=FileHandoffResult, summary="Pass a file between agents")
def openclaw_file_handoff(
    payload: FileHandoffRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Copy a working file from one agent's folder to another's (file flows in a team)."""
    return service.handoff_file(
        from_agent_id=payload.from_agent_id,
        to_agent_id=payload.to_agent_id,
        file_name=payload.file_name,
    )


@router.post("/web-search/enable", response_model=WebSearchResult, summary="Enable free web search")
def openclaw_enable_web_search(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    """Turn on free DuckDuckGo web search (no API key) so agents can search the net."""
    return service.enable_web_search()


@router.get("/channels", response_model=ChannelsResponse, summary="List chat channels")
def openclaw_channels(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.list_channels()


@router.post("/channels/add", response_model=ChannelActionResult, summary="Connect a channel")
def openclaw_channels_add(
    payload: AddChannelRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Connect a messaging platform (Discord, Telegram, Slack, …) with a token."""
    return service.add_channel(channel=payload.channel, token=payload.token)


@router.get("/agents", response_model=OpenClawAgentsResponse, summary="List OpenClaw agents")
def openclaw_agents(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    return service.list_agents()


@router.post("/agents", response_model=OpenClawAgentResult, summary="Create an OpenClaw agent")
def openclaw_agent_create(
    payload: CreateOpenClawAgentRequest,
    service: OpenClawService = Depends(get_openclaw_service),
    session: Session = Depends(get_session),
    settings: Settings = Depends(get_app_settings),
) -> dict:
    """Create an isolated OpenClaw agent (Gemma 4 by default) with optional
    instructions and identity — it then runs inside OpenClaw. When AFM is
    configured, the new agent's folder is created inside the AFM root."""
    from ..services.afm_service import AFMService

    afm = AFMService(session, settings)
    workspace = str(afm.agent_target_dir(payload.name)) if afm.is_configured() else None
    return service.create_agent(
        name=payload.name, model=payload.model,
        instructions=payload.instructions, emoji=payload.emoji,
        workspace=workspace,
    )


@router.delete("/agents/{agent_id}", response_model=AgentActionResult, summary="Delete an OpenClaw agent")
def openclaw_agent_delete(
    agent_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.delete_agent(agent_id)


@router.post("/agents/{agent_id}/chat", response_model=AgentChatResult, summary="Chat with an OpenClaw agent")
def openclaw_agent_chat(
    agent_id: str,
    payload: AgentChatRequest,
    service: OpenClawService = Depends(get_openclaw_service),
    session: Session = Depends(get_session),
    settings: Settings = Depends(get_app_settings),
) -> dict:
    """Run one agent turn natively (no dashboard). Pass a stable session_id to keep
    conversation continuity. Can take a while on local models. If the local
    knowledge base has relevant content, it is prepended as context (RAG)."""
    knowledge = KnowledgeService(session, settings, ChromaService(settings))
    message = knowledge.context_block(payload.message) + payload.message

    task = (payload.activity_task or "").strip() or "Answering a chat"
    kind = payload.activity_kind if payload.activity_kind in ("chat", "team") else "chat"
    with track(agent_id=f"openclaw:{agent_id}", agent_name=agent_id, kind=kind, task=task) as outcome:
        result = service.agent_chat(
            agent_id=agent_id, message=message, session_id=payload.session_id
        )
        outcome["ok"] = bool(result.get("ok"))
        if not outcome["ok"]:
            outcome["note"] = result.get("message") or ""
    return result


@router.get("/documents", response_model=AgentDocumentsResponse, summary="Agent work files, grouped")
def openclaw_agent_documents(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    """The files agents created in their workspaces (reports, drafts, team
    working files) — excluding config/memory/skills. Powers the Documents tab."""
    return service.list_agent_documents()


@router.get("/documents/content", response_model=AgentDocumentContent, summary="Read an agent work file")
def openclaw_agent_document_content(
    agent_id: str,
    path: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Bounded text preview of a work file (path is relative to the agent's
    workspace; traversal and config files are refused)."""
    return service.read_agent_document(agent_id, path)


@router.delete("/documents/content", response_model=AgentActionResult, summary="Delete an agent work file")
def openclaw_agent_document_delete(
    agent_id: str,
    path: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.delete_agent_document(agent_id, path)


@router.get("/chats", response_model=AllChatsResponse, summary="All conversations, grouped by agent")
def openclaw_all_chats(service: OpenClawService = Depends(get_openclaw_service)) -> dict:
    """Every saved conversation of every agent, with where it came from
    (Evano app, a team run, Discord, the dashboard, …). Powers the Chats tab."""
    return service.list_all_chats()


@router.get("/agents/{agent_id}/discord", response_model=AgentDiscordStatus, summary="Agent's Discord link")
def openclaw_agent_discord_status(
    agent_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Is this agent connected to Discord as its own bot?"""
    return service.agent_discord_status(agent_id)


@router.post("/agents/{agent_id}/discord", response_model=AgentDiscordResult, summary="Connect agent to Discord")
def openclaw_agent_discord_connect(
    agent_id: str,
    payload: ConnectDiscordRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Make this agent a Discord bot: register its bot token as the agent's own
    channel account, bind it, and reload the gateway. The token is never logged."""
    return service.connect_agent_discord(agent_id=agent_id, bot_token=payload.bot_token)


@router.delete("/agents/{agent_id}/discord", response_model=AgentDiscordResult, summary="Disconnect agent from Discord")
def openclaw_agent_discord_disconnect(
    agent_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.disconnect_agent_discord(agent_id)


@router.get("/agents/{agent_id}/sessions", response_model=SessionsResponse, summary="Conversation history")
def openclaw_agent_sessions(
    agent_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """List saved conversations for an agent (persisted by OpenClaw on disk)."""
    return service.list_sessions(agent_id)


@router.get("/agents/{agent_id}/sessions/{session_id}", response_model=SessionDetail, summary="Load a conversation")
def openclaw_agent_session(
    agent_id: str,
    session_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.get_session(agent_id, session_id)


@router.delete("/agents/{agent_id}/sessions/{session_id}", response_model=AgentActionResult, summary="Delete a conversation")
def openclaw_agent_session_delete(
    agent_id: str,
    session_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.delete_session(agent_id, session_id)


@router.post("/agents/{agent_id}/sessions/clear", response_model=ClearSessionsResult, summary="Clear conversation history")
def openclaw_agent_sessions_clear(
    agent_id: str,
    payload: ClearSessionsRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Delete all conversations (or only those older than N days) to free disk."""
    return service.clear_sessions(agent_id, older_than_days=payload.older_than_days)


@router.get("/agents/{agent_id}/files", response_model=AgentFilesResponse, summary="Agent config files")
def openclaw_agent_files(
    agent_id: str,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    """Read the agent's workspace config files (IDENTITY/AGENTS/SOUL/…)."""
    return service.get_agent_files(agent_id)


@router.put("/agents/{agent_id}/files/{name}", response_model=AgentActionResult, summary="Save an agent config file")
def openclaw_agent_file_save(
    agent_id: str,
    name: str,
    payload: SaveAgentFileRequest,
    service: OpenClawService = Depends(get_openclaw_service),
) -> dict:
    return service.save_agent_file(agent_id=agent_id, name=name, content=payload.content)

File: apps/desktop/src/renderer/src/lib/api/endpoints/openclaw.ts
/** openclaw endpoints. */
import { requestJson } from "../http";
import type {
  OpenClawStatus,
  OpenClawInstallStatus,
  OpenClawConfigResult,
  GatewayActionResult,
  DashboardUrlResponse,
  WebSearchResult,
  ChannelsResponse,
  ChannelActionResult,
  OpenClawAgentsResponse,
  OpenClawAgentResult,
  AgentActionResult,
  AgentChatResult,
  AgentDiscordResult,
  AgentDiscordStatus,
  AgentDocumentContent,
  AgentDocumentsResponse,
  AgentFilesResponse,
  AllChatsResponse,
  SessionsResponse,
  SessionDetail,
  ClearSessionsResult,
  FileHandoffResult,
} from "../types";

export function getOpenClawStatus(): Promise<OpenClawStatus> {
  return requestJson<OpenClawStatus>("/openclaw/status");
}

/** `POST /openclaw/file-handoff` — copy a working file from one agent to another. */
export function handoffOpenClawFile(body: {
  from_agent_id: string;
  to_agent_id: string;
  file_name: string;
}): Promise<FileHandoffResult> {
  return requestJson<FileHandoffResult>("/openclaw/file-handoff", {
    method: "POST",
    body,
    timeoutMs: 30_000,
  });
}

/** `POST /openclaw/install` — start installing OpenClaw via npm (non-blocking). */
export function installOpenClaw(): Promise<OpenClawInstallStatus> {
  return requestJson<OpenClawInstallStatus>("/openclaw/install", { method: "POST" });
}

/** `GET /openclaw/install/status` — install progress. */
export function getOpenClawInstallStatus(): Promise<OpenClawInstallStatus> {
  return requestJson<OpenClawInstallStatus>("/openclaw/install/status");
}

/** `POST /openclaw/config` — configure OpenClaw (free Gemma 4, or API key). */
export function writeOpenClawConfig(body: {
  mode: "free" | "api";
  model?: string;
  provider?: string;
  api_key?: string;
  base_url?: string;
}): Promise<OpenClawConfigResult> {
  // Onboarding can take a while (validates/sets up the workspace).
  return requestJson<OpenClawConfigResult>("/openclaw/config", {
    method: "POST",
    body,
    timeoutMs: 200_000,
  });
}

/** `POST /openclaw/gateway/start` — start the OpenClaw gateway. */
export function startOpenClawGateway(): Promise<GatewayActionResult> {
  return requestJson<GatewayActionResult>("/openclaw/gateway/start", { method: "POST", timeoutMs: 70_000 });
}

/** `POST /openclaw/gateway/stop` — stop the OpenClaw gateway. */
export function stopOpenClawGateway(): Promise<GatewayActionResult> {
  return requestJson<GatewayActionResult>("/openclaw/gateway/stop", { method: "POST", timeoutMs: 70_000 });
}

/** `GET /openclaw/dashboard` — token-authenticated dashboard URL. */
export function getOpenClawDashboardUrl(): Promise<DashboardUrlResponse> {
  return requestJson<DashboardUrlResponse>("/openclaw/dashboard", { timeoutMs: 70_000 });
}

/** `POST /openclaw/web-search/enable` — turn on free DuckDuckGo web search. */
export function enableOpenClawWebSearch(): Promise<WebSearchResult> {
  return requestJson<WebSearchResult>("/openclaw/web-search/enable", { method: "POST", timeoutMs: 70_000 });
}

/** `GET /openclaw/channels` — messaging platforms + connection state. */
export function getOpenClawChannels(): Promise<ChannelsResponse> {
  return requestJson<ChannelsResponse>("/openclaw/channels", { timeoutMs: 40_000 });
}

/** `POST /openclaw/channels/add` — connect a channel with a token. */
export function addOpenClawChannel(body: { channel: string; token: string }): Promise<ChannelActionResult> {
  return requestJson<ChannelActionResult>("/openclaw/channels/add", {
    method: "POST",
    body,
    timeoutMs: 70_000,
  });
}

/** `GET /openclaw/agents` — agents that run inside OpenClaw. */
export function getOpenClawAgents(): Promise<OpenClawAgentsResponse> {
  return requestJson<OpenClawAgentsResponse>("/openclaw/agents", { timeoutMs: 40_000 });
}

/** `POST /openclaw/agents` — create an OpenClaw agent (Gemma 4 by default). */
export function createOpenClawAgent(body: {
  name: string;
  model?: string;
  instructions?: string;
  emoji?: string;
}): Promise<OpenClawAgentResult> {
  return requestJson<OpenClawAgentResult>("/openclaw/agents", { method: "POST", body, timeoutMs: 90_000 });
}

/** `DELETE /openclaw/agents/{id}` — delete an OpenClaw agent. */
export function deleteOpenClawAgent(id: string): Promise<AgentActionResult> {
  return requestJson<AgentActionResult>(`/openclaw/agents/${encodeURIComponent(id)}`, {
    method: "DELETE",
    timeoutMs: 70_000,
  });
}

/** `POST /openclaw/agents/{id}/chat` — one native agent turn (can be slow). A
 *  stable sessionId keeps conversation continuity (the agent remembers). The
 *  optional activity label/kind is what the Office view shows while it runs. */
export function chatOpenClawAgent(
  id: string,
  message: string,
  sessionId?: string,
  activity?: { task: string; kind: "chat" | "team" },
): Promise<AgentChatResult> {
  return requestJson<AgentChatResult>(`/openclaw/agents/${encodeURIComponent(id)}/chat`, {
    method: "POST",
    body: {
      message,
      session_id: sessionId,
      activity_task: activity?.task,
      activity_kind: activity?.kind,
    },
    timeoutMs: 240_000,
  });
}

/** `GET /openclaw/agents/{id}/discord` — is this agent live as its own Discord bot? */
export function getAgentDiscordStatus(id: string): Promise<AgentDiscordStatus> {
  return requestJson<AgentDiscordStatus>(`/openclaw/agents/${encodeURIComponent(id)}/discord`, {
    timeoutMs: 40_000,
  });
}

/** `POST /openclaw/agents/{id}/discord` — register the bot token + bind + reload gateway. */
export function connectAgentDiscord(id: string, botToken: string): Promise<AgentDiscordResult> {
  return requestJson<AgentDiscordResult>(`/openclaw/agents/${encodeURIComponent(id)}/discord`, {
    method: "POST",
    body: { bot_token: botToken },
    timeoutMs: 120_000, // may restart the gateway
  });
}

/** `DELETE /openclaw/agents/{id}/discord` — unlink the bot and remove its account. */
export function disconnectAgentDiscord(id: string): Promise<AgentDiscordResult> {
  return requestJson<AgentDiscordResult>(`/openclaw/agents/${encodeURIComponent(id)}/discord`, {
    method: "DELETE",
    timeoutMs: 120_000,
  });
}

/** `GET /openclaw/agents/{id}/files` — the agent's config files (IDENTITY/AGENTS/…). */
export function getOpenClawAgentFiles(id: string): Promise<AgentFilesResponse> {
  return requestJson<AgentFilesResponse>(`/openclaw/agents/${encodeURIComponent(id)}/files`, {
    timeoutMs: 40_000,
  });
}

/** `PUT /openclaw/agents/{id}/files/{name}` — save a config file. */
export function saveOpenClawAgentFile(id: string, name: string, content: string): Promise<AgentActionResult> {
  return requestJson<AgentActionResult>(
    `/openclaw/agents/${encodeURIComponent(id)}/files/${encodeURIComponent(name)}`,
    { method: "PUT", body: { content }, timeoutMs: 40_000 },
  );
}

/** `GET /openclaw/documents` — agent work files, grouped by agent. */
export function getAgentDocuments(): Promise<AgentDocumentsResponse> {
  return requestJson<AgentDocumentsResponse>("/openclaw/documents", { timeoutMs: 40_000 });
}

/** `GET /openclaw/documents/content` — bounded text preview of a work file. */
export function getAgentDocumentContent(agentId: string, path: string): Promise<AgentDocumentContent> {
  const qs = `agent_id=${encodeURIComponent(agentId)}&path=${encodeURIComponent(path)}`;
  return requestJson<AgentDocumentContent>(`/openclaw/documents/content?${qs}`, { timeoutMs: 40_000 });
}

/** `DELETE /openclaw/documents/content` — delete an agent work file. */
export function deleteAgentDocument(agentId: string, path: string): Promise<AgentActionResult> {
  const qs = `agent_id=${encodeURIComponent(agentId)}&path=${encodeURIComponent(path)}`;
  return requestJson<AgentActionResult>(`/openclaw/documents/content?${qs}`, {
    method: "DELETE",
    timeoutMs: 40_000,
  });
}

/** `GET /openclaw/chats` — every conversation of every agent, with origins. */
export function getAllChats(): Promise<AllChatsResponse> {
  return requestJson<AllChatsResponse>("/openclaw/chats", { timeoutMs: 40_000 });
}

/** `GET /openclaw/agents/{id}/sessions` — saved conversation history. */
export function getOpenClawSessions(id: string): Promise<SessionsResponse> {
  return requestJson<SessionsResponse>(`/openclaw/agents/${encodeURIComponent(id)}/sessions`, {
    timeoutMs: 40_000,
  });
}

/** `GET /openclaw/agents/{id}/sessions/{sid}` — load a past conversation. */
export function getOpenClawSession(id: string, sessionId: string): Promise<SessionDetail> {
  return requestJson<SessionDetail>(
    `/openclaw/agents/${encodeURIComponent(id)}/sessions/${encodeURIComponent(sessionId)}`,
    { timeoutMs: 40_000 },
  );
}

/** `DELETE /openclaw/agents/{id}/sessions/{sid}` — delete one conversation. */
export function deleteOpenClawSession(id: string, sessionId: string): Promise<AgentActionResult> {
  return requestJson<AgentActionResult>(
    `/openclaw/agents/${encodeURIComponent(id)}/sessions/${encodeURIComponent(sessionId)}`,
    { method: "DELETE", timeoutMs: 40_000 },
  );
}

/** `POST /openclaw/agents/{id}/sessions/clear` — clear history (optionally only older than N days). */
export function clearOpenClawSessions(id: string, olderThanDays?: number): Promise<ClearSessionsResult> {
  return requestJson<ClearSessionsResult>(`/openclaw/agents/${encodeURIComponent(id)}/sessions/clear`, {
    method: "POST",
    body: { older_than_days: olderThanDays ?? null },
    timeoutMs: 40_000,
  });
}

/** `GET /settings` — all app settings. */
